File: config/items/slider.py
from flask_babel import gettext
from config.items.default import Item


class Slider(Item):
    """
    Slider item subclass
    """

    TYPE = "slider"
    VISIBLE = True
    NAME = gettext("Slider")
    DESCRIPTION = gettext("Slide to set value")
    VALUE = [50, 80]

    @property
    def config(self):
        return {
            self._LABEL: self.NAME,
            self._SUFFIX: "%",
            self._MIN: 0,
            self._MAX: 100,
            self._STEP: 5,
            self._SMOOTH: False,
            self._RANGE: False,
            self._DISABLED: False
        }

    @property
    def edit_config(self):
        from config.items.input import Input
        from config.items.toggle import Toggle

        return {
            self._LABEL: Input().make_object(value=self.config[self._LABEL], label=gettext("Label")),
            self._SUFFIX: Input().make_object(value=self.config[self._SUFFIX], label=gettext("Suffix")),
            self._MIN: Input().make_object(value=self.config[self._MIN], label=gettext("Minimal value")),
            self._MAX: Input().make_object(value=self.config[self._MAX], label=gettext("Maximal value")),
            self._STEP: Input().make_object(value=self.config[self._STEP], label=gettext("Step value")),
            self._SMOOTH: Toggle().make_object(value=self.config[self._SMOOTH], label=gettext("Smooth")),
            self._RANGE: Toggle().make_object(value=self.config[self._RANGE], label=gettext("Range")),
            self._DISABLED: Toggle().make_object(value=self.config[self._DISABLED], label=gettext("Disabled"))
        }

File: static/js/items/slider.js
"use strict";

class itemSlider {
  constructor() {
    this.type = "slider";
  }

  setup(object) {}

  valueReceive(object, self, value, config) {
    let key;

    if (config.range === true) {
      key = "values"
      $(self).find(".modal-slider-value").text(value[0] + " - " + value[1]);
    }
    else {
      key = "value"
      $(self).find("label").text(config.label);
      $(self).find(".modal-slider-value").text(value);
    }

    $(self).find(".slider").slider(key, value);
  }

  valueTransmit(object) {
    let initConfig = store(object, "config");
    let initValue = store(object, "value");
    object.find('.slider').slider({
        slide: function(event, ui) {
          let config = store(object, "config");
          let value;
          if (config.range === true) {
            value = $(this).slider("values")
            value[ui.handleIndex] = ui.value
            object.find(".modal-slider-value").text(value[0] + " - " + value[1]);
          }
          else {
            value = ui.value;
            object.find(".modal-slider-value").text(value);
          }

          if (config["smooth"] === true) {
            object.find(".modal-slider-right").css("opacity", 1);
            store(object, "value", value).trigger("value-transmit");
          }
          else {
            object.find(".modal-slider-right").css("opacity", 0.6);
          }
        },
        stop: function(event, ui) {
          let config = store(object, "config");

          if (!(config["smooth"])) {
            // Send only if slider changed value
            object.find(".modal-slider-right").css("opacity", 1);
            let value;
            if (config.range === true) {
              value = $(this).slider("values");
              object.find(".modal-slider-value").text(value[0] + " - " + value[1]);
            }
            else {
              value = ui.value;
              object.find(".modal-slider-value").text(value);
            }
            store(object, "value", value).trigger("value-transmit");
          }
        },
        animate: "fast",
        min: initConfig.min,
        max: initConfig.max,
        value:  initConfig.range ? null : (Array.isArray(initValue) ? initValue[0] : initValue),
        values: initConfig.range ? (Array.isArray(initValue) ? initValue    : [initValue, initValue]) : null,
        range: initConfig.range ? true : "min",
        step: initConfig.step,
        disabled: initConfig.disabled
    });
  }

  configReceive(object, self, value, config) {
    let range = config.range;
    if (range !== true) range = "min";

    $(self).find(".modal-slider-label").text(config.label);
    $(self).find(".modal-slider-suffix").text(" " + config.suffix);
    let options = {
      range: range,
      min: config["min"],
      max: config["max"],
      step: config["step"],
      disabled: config["disabled"]};

    $(self).find(".slider").slider("option", options);

    if (range === true) {
      if (!(Array.isArray(value))) {
        let sliderValue = $(self).find(".slider").slider("values");
        if(!(sliderValue) || (sliderValue[0] === 0 && sliderValue[1] === 0)) {
          $(self).find(".slider").slider("option", "values", [value, value]);
        }
      }
      else {
        $(self).find(".slider").slider("option", "values", value);
      }
    }
    else {
      if (Array.isArray(value)) {
        if(!($(self).find(".slider").slider("value"))) {
          $(self).find(".slider").slider("option", "value", value[0]);
        }
      }
    }
  }
}